# Import Packages

In [96]:
#Load Packages
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import os as os
import requests
import io
import urllib.request
from sklearn.metrics import mean_squared_error
from scipy.io import arff
%matplotlib inline 

In [97]:
%pwd

'C:\\Users\\JM025575\\Predictive Models Class\\data'

In [98]:
cd /Users/JM025575/Predictive Models Class/data

C:\Users\JM025575\Predictive Models Class\data


# Load Data

In [99]:
attrib = pd.read_csv('attributes.csv', delim_whitespace = True)

In [100]:
url = 'http://archive.ics.uci.edu/ml/machine-learning-databases/communities/communities.data'
crime = pd.read_csv(url, names = attrib['attributes'])
crime.head()

state county community        communityname  fold  population  \
0      8      ?         ?         Lakewoodcity     1        0.19   
1     53      ?         ?          Tukwilacity     1        0.00   
2     24      ?         ?         Aberdeentown     1        0.00   
3     34      5     81440  Willingborotownship     1        0.04   
4     42     95      6096    Bethlehemtownship     1        0.01   

   householdsize  racepctblack  racePctWhite  racePctAsian  \
0           0.33          0.02          0.90          0.12   
1           0.16          0.12          0.74          0.45   
2           0.42          0.49          0.56          0.17   
3           0.77          1.00          0.08          0.12   
4           0.55          0.02          0.95          0.09   

          ...           LandArea  PopDens  PctUsePubTrans  PolicCars  \
0         ...               0.12     0.26            0.20       0.06   
1         ...               0.02     0.12            0.45          ?   
2         ...               0.01     0.21            0.02          ?   
3         ...               0.02     0.39            0.28          ?   
4         ...               0.04     0.09            0.02          ?   

   PolicOperBudg  LemasPctPolicOnPatr  LemasGangUnitDeploy  \
0           0.04                  0.9                  0.5   
1              ?                    ?                    ?   
2              ?                    ?                    ?   
3              ?                    ?                    ?   
4              ?                    ?                    ?   

   LemasPctOfficDrugUn  PolicBudgPerPop  ViolentCrimesPerPop  
0                 0.32             0.14                 0.20  
1                 0.00                ?                 0.67  
2                 0.00                ?                 0.43  
3                 0.00                ?                 0.12  
4                 0.00                ?                 0.03  

[5 rows x 128 columns]

In [101]:
crime.columns.tolist()

['state',
 'county',
 'community',
 'communityname',
 'fold',
 'population',
 'householdsize',
 'racepctblack',
 'racePctWhite',
 'racePctAsian',
 'racePctHisp',
 'agePct12t21',
 'agePct12t29',
 'agePct16t24',
 'agePct65up',
 'numbUrban',
 'pctUrban',
 'medIncome',
 'pctWWage',
 'pctWFarmSelf',
 'pctWInvInc',
 'pctWSocSec',
 'pctWPubAsst',
 'pctWRetire',
 'medFamInc',
 'perCapInc',
 'whitePerCap',
 'blackPerCap',
 'indianPerCap',
 'AsianPerCap',
 'OtherPerCap',
 'HispPerCap',
 'NumUnderPov',
 'PctPopUnderPov',
 'PctLess9thGrade',
 'PctNotHSGrad',
 'PctBSorMore',
 'PctUnemployed',
 'PctEmploy',
 'PctEmplManu',
 'PctEmplProfServ',
 'PctOccupManu',
 'PctOccupMgmtProf',
 'MalePctDivorce',
 'MalePctNevMarr',
 'FemalePctDiv',
 'TotalPctDiv',
 'PersPerFam',
 'PctFam2Par',
 'PctKids2Par',
 'PctYoungKids2Par',
 'PctTeen2Par',
 'PctWorkMomYoungKids',
 'PctWorkMom',
 'NumIlleg',
 'PctIlleg',
 'NumImmig',
 'PctImmigRecent',
 'PctImmigRec5',
 'PctImmigRec8',
 'PctImmigRec10',
 'PctRecentImmig',
 'P

In [102]:
#getting rid of features I do not think have predictive power
crime = crime.drop(['state','county','community','communityname','fold'], axis=1)

In [103]:
crime.head()

population  householdsize  racepctblack  racePctWhite  racePctAsian  \
0        0.19           0.33          0.02          0.90          0.12   
1        0.00           0.16          0.12          0.74          0.45   
2        0.00           0.42          0.49          0.56          0.17   
3        0.04           0.77          1.00          0.08          0.12   
4        0.01           0.55          0.02          0.95          0.09   

   racePctHisp  agePct12t21  agePct12t29  agePct16t24  agePct65up  \
0         0.17         0.34         0.47         0.29        0.32   
1         0.07         0.26         0.59         0.35        0.27   
2         0.04         0.39         0.47         0.28        0.32   
3         0.10         0.51         0.50         0.34        0.21   
4         0.05         0.38         0.38         0.23        0.36   

          ...           LandArea  PopDens  PctUsePubTrans  PolicCars  \
0         ...               0.12     0.26            0.20       0.06   
1         ...               0.02     0.12            0.45          ?   
2         ...               0.01     0.21            0.02          ?   
3         ...               0.02     0.39            0.28          ?   
4         ...               0.04     0.09            0.02          ?   

   PolicOperBudg  LemasPctPolicOnPatr  LemasGangUnitDeploy  \
0           0.04                  0.9                  0.5   
1              ?                    ?                    ?   
2              ?                    ?                    ?   
3              ?                    ?                    ?   
4              ?                    ?                    ?   

   LemasPctOfficDrugUn  PolicBudgPerPop  ViolentCrimesPerPop  
0                 0.32             0.14                 0.20  
1                 0.00                ?                 0.67  
2                 0.00                ?                 0.43  
3                 0.00                ?                 0.12  
4                 0.00                ?                 0.03  

[5 rows x 123 columns]

In [104]:
crime.shape

(1994, 123)

In [105]:
#Reviewing basic statistics on my target variable
crime['ViolentCrimesPerPop'].describe()

count    1994.000000
mean        0.237979
std         0.232985
min         0.000000
25%         0.070000
50%         0.150000
75%         0.330000
max         1.000000
Name: ViolentCrimesPerPop, dtype: float64

## Data Pre-Processing

In [106]:
#Replacing the '?' with nan and looking to identify features that have null values
crime = crime.replace('?', np.nan)
feat_miss = crime.columns[crime.isnull().any()]

In [107]:
print(feat_miss)
feat_miss.shape #23 of the features have missing values

Index(['OtherPerCap', 'LemasSwornFT', 'LemasSwFTPerPop', 'LemasSwFTFieldOps',
       'LemasSwFTFieldPerPop', 'LemasTotalReq', 'LemasTotReqPerPop',
       'PolicReqPerOffic', 'PolicPerPop', 'RacialMatchCommPol',
       'PctPolicWhite', 'PctPolicBlack', 'PctPolicHisp', 'PctPolicAsian',
       'PctPolicMinor', 'OfficAssgnDrugUnits', 'NumKindsDrugsSeiz',
       'PolicAveOTWorked', 'PolicCars', 'PolicOperBudg', 'LemasPctPolicOnPatr',
       'LemasGangUnitDeploy', 'PolicBudgPerPop'],
      dtype='object')


(23,)

In [108]:
#removing features with null values
crime = crime.dropna(axis=1)
print(crime.shape)
crime.head() #go from 123 to 100 features

(1994, 100)


population  householdsize  racepctblack  racePctWhite  racePctAsian  \
0        0.19           0.33          0.02          0.90          0.12   
1        0.00           0.16          0.12          0.74          0.45   
2        0.00           0.42          0.49          0.56          0.17   
3        0.04           0.77          1.00          0.08          0.12   
4        0.01           0.55          0.02          0.95          0.09   

   racePctHisp  agePct12t21  agePct12t29  agePct16t24  agePct65up  \
0         0.17         0.34         0.47         0.29        0.32   
1         0.07         0.26         0.59         0.35        0.27   
2         0.04         0.39         0.47         0.28        0.32   
3         0.10         0.51         0.50         0.34        0.21   
4         0.05         0.38         0.38         0.23        0.36   

          ...           PctForeignBorn  PctBornSameState  PctSameHouse85  \
0         ...                     0.12              0.42            0.50   
1         ...                     0.21              0.50            0.34   
2         ...                     0.14              0.49            0.54   
3         ...                     0.19              0.30            0.73   
4         ...                     0.11              0.72            0.64   

   PctSameCity85  PctSameState85  LandArea  PopDens  PctUsePubTrans  \
0           0.51            0.64      0.12     0.26            0.20   
1           0.60            0.52      0.02     0.12            0.45   
2           0.67            0.56      0.01     0.21            0.02   
3           0.64            0.65      0.02     0.39            0.28   
4           0.61            0.53      0.04     0.09            0.02   

   LemasPctOfficDrugUn  ViolentCrimesPerPop  
0                 0.32                 0.20  
1                 0.00                 0.67  
2                 0.00                 0.43  
3                 0.00                 0.12  
4                 0.00                 0.03  

[5 rows x 100 columns]

In [109]:
#Define the features and target variable, 'Violent Crimes per Pop'
crime.target=crime['ViolentCrimesPerPop'] 
crime.features=crime.drop(['ViolentCrimesPerPop'], axis=1)                         
print(crime.target.shape)
print(crime.features.shape)
crime.head()

(1994,)
(1994, 99)


C:\Users\JM025575\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:2: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  
C:\Users\JM025575\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:3: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  This is separate from the ipykernel package so we can avoid doing imports until


population  householdsize  racepctblack  racePctWhite  racePctAsian  \
0        0.19           0.33          0.02          0.90          0.12   
1        0.00           0.16          0.12          0.74          0.45   
2        0.00           0.42          0.49          0.56          0.17   
3        0.04           0.77          1.00          0.08          0.12   
4        0.01           0.55          0.02          0.95          0.09   

   racePctHisp  agePct12t21  agePct12t29  agePct16t24  agePct65up  \
0         0.17         0.34         0.47         0.29        0.32   
1         0.07         0.26         0.59         0.35        0.27   
2         0.04         0.39         0.47         0.28        0.32   
3         0.10         0.51         0.50         0.34        0.21   
4         0.05         0.38         0.38         0.23        0.36   

          ...           PctForeignBorn  PctBornSameState  PctSameHouse85  \
0         ...                     0.12              0.42            0.50   
1         ...                     0.21              0.50            0.34   
2         ...                     0.14              0.49            0.54   
3         ...                     0.19              0.30            0.73   
4         ...                     0.11              0.72            0.64   

   PctSameCity85  PctSameState85  LandArea  PopDens  PctUsePubTrans  \
0           0.51            0.64      0.12     0.26            0.20   
1           0.60            0.52      0.02     0.12            0.45   
2           0.67            0.56      0.01     0.21            0.02   
3           0.64            0.65      0.02     0.39            0.28   
4           0.61            0.53      0.04     0.09            0.02   

   LemasPctOfficDrugUn  ViolentCrimesPerPop  
0                 0.32                 0.20  
1                 0.00                 0.67  
2                 0.00                 0.43  
3                 0.00                 0.12  
4                 0.00                 0.03  

[5 rows x 100 columns]

# Ridge Regression

In [110]:
# Ridge Regression package
from sklearn.linear_model import Ridge

## Basic Ridge Model

In [134]:
# fit a ridge regression model to the data
model_RG = Ridge(alpha = 1.0)
model_RG.fit(crime.features, crime.target)
print(model_RG)
# make predictions
expected_RG = crime.target
predicted_RG= model_RG.predict(crime.features)
# summarize the fit of the model
print("Coef", model_RG.intercept_, model_RG.coef_)
print("MSE", mean_squared_error(expected_RG, predicted_RG))

Ridge(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)
Coef 0.5135351816663232 [-3.38418572e-02  2.43357321e-02  1.87027498e-01 -5.23197422e-02
 -1.67677668e-02  4.72345392e-02  4.14437009e-02 -1.74875816e-01
 -4.36574600e-02  4.32258031e-02 -6.52537224e-02  3.97657237e-02
  3.39623356e-02 -1.24624046e-01  3.67306261e-02 -1.47755710e-01
  5.96513815e-02  1.49875553e-02 -8.11220058e-02  6.33875794e-02
 -2.06677307e-02 -1.62903535e-01 -2.00122714e-02 -3.05253966e-02
  2.47912310e-02  4.52935960e-02  3.43761924e-02 -1.38473937e-01
 -9.01636921e-02  5.05364563e-02  4.69968259e-02 -1.78613967e-02
  1.53497534e-01 -5.54816676e-02 -1.67671551e-02  6.17033463e-02
  5.02487862e-02  1.58507615e-01  1.36778968e-01 -8.67717489e-02
 -2.85580569e-02  2.56240659e-02 -6.40886760e-02 -1.81783063e-01
 -4.25160303e-02 -8.68274480e-03  3.26748401e-02 -1.47477326e-01
 -5.88770547e-02  1.38820453e-01 -9.62680028e-02  2.36638138e-02
 

The basic Ridge Regression model has a MSE of 0.0168 which seems very low considering I have not adjusted the model yet.

In [135]:
#pd.DataFrame(np.transpose(model_RG.coef_))

In [136]:
#Adjusting the alpha in my model from the default of 1 to a 5.
# fit a ridge regression model to the data
model_RG = Ridge(alpha = 5.0)
model_RG.fit(crime.features, crime.target)
print(model_RG)
# make predictions
expected_RG = crime.target
predicted_RG= model_RG.predict(crime.features)
# summarize the fit of the model
print("Coef", model_RG.intercept_, model_RG.coef_)
print("MSE", mean_squared_error(expected_RG, predicted_RG))

Ridge(alpha=5.0, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)
Coef 0.45307973719791744 [-1.74029508e-02  2.19538588e-02  1.56808585e-01 -7.84171774e-02
 -2.25143656e-02  2.67662320e-02  3.17058614e-03 -7.62612368e-02
 -2.24813238e-02  3.29048250e-02 -2.38156131e-02  3.77737574e-02
  2.38288029e-02 -5.94761486e-02  2.37566650e-02 -9.57319569e-02
  3.47141852e-02  2.29747913e-02 -6.66644488e-02  1.44559973e-02
 -1.99716872e-02 -4.99976600e-02 -1.61532136e-02 -2.64146695e-02
  2.49949205e-02  3.69949464e-02  4.13048760e-03 -8.15859840e-02
 -5.26527986e-02  2.73622797e-02  6.94197550e-03 -2.69056556e-02
  5.72172454e-02 -3.93827068e-02 -8.65758659e-03  3.12265207e-02
  9.05615658e-03  9.25021535e-02  6.40847500e-02 -2.70179709e-02
  1.33994221e-02  2.35692417e-02 -6.83992579e-02 -1.14110877e-01
 -5.69656535e-02 -2.14757441e-02  2.76415751e-03 -8.70753676e-02
 -6.11495909e-03  1.45608971e-01 -4.04414090e-02  1.56746422e-02


My second Ridge Regression model I adjusted my alpha, or error, up to 5.0.  This caused my MSE to go up slightly to 0.0173.

In [137]:
#Adjusting the alpha in my model from 5 to 0.5.
# fit a ridge regression model to the data
model_RG = Ridge(alpha = 0.5)
model_RG.fit(crime.features, crime.target)
print(model_RG)
# make predictions
expected_RG = crime.target
predicted_RG= model_RG.predict(crime.features)
# summarize the fit of the model
print("Coef", model_RG.intercept_, model_RG.coef_)
print("MSE", mean_squared_error(expected_RG, predicted_RG))

Ridge(alpha=0.5, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)
Coef 0.5380523806944332 [-0.02769673  0.01625281  0.19643194 -0.0451303  -0.01425859  0.0521199
  0.06057172 -0.21382197 -0.05803366  0.04034612 -0.08664645  0.04051655
  0.01305794 -0.1509612   0.04035378 -0.16043553  0.06859303  0.01097773
 -0.08269686  0.09955208 -0.00199537 -0.21704958 -0.02167977 -0.03175173
  0.02377664  0.04680228  0.06045067 -0.15275291 -0.1009773   0.05966999
  0.05653957 -0.01113828  0.19018832 -0.06008011 -0.02044392  0.0694102
  0.06656798  0.18757915  0.16833297 -0.09702482 -0.05716057  0.00793619
 -0.04924118 -0.21818419 -0.03682284 -0.00612913  0.04323305 -0.16456274
 -0.08373667  0.13227816 -0.11450864  0.02514657 -0.00576413 -0.02334321
  0.01093488 -0.02884964 -0.0337178   0.10062989 -0.0085738  -0.02552047
 -0.1308538  -0.05910759 -0.07000419  0.30045838 -0.1052186  -0.09454653
 -0.17236221  0.19000032  0.08372958  0.02041

My third Ridge Regression model, I adjusted my alpha down to 0.5 to see if that would cause my MSE to go back down.  It did cause it to go down, even slightly lower than the basic model, down to 0.0167.

## Grid Search to narrow down alpha and Cross Validation 10 times

In [141]:
# use a full grid over several parameters
from sklearn.model_selection import GridSearchCV
param_grid = {"alpha": [0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]}  #different penalties I am playing with

# run grid search
grid_search = GridSearchCV(model_RG, param_grid=param_grid,n_jobs=-1, cv = 10)
grid_search.fit(crime.features, crime.target)
print("Best", grid_search.best_params_) 

Best {'alpha': 0.9}


Using a grid search, I ran 8 different alpha values through my model.  I use the grid search because it is a more efficient way of trying to find what my most ideal alpha value is rather than plugging one in at a time and seeing how my MSE adjusts.  I also used cross validation with the grid search cross validating 10 times.  Cross validation allows my model to be run a specific number of times, 10 in this case, each time taking different samples of data to run the model against for my various alphas I am using in the grid search.  Cross validation is a good way to see if we are overfitting our model. My result was that my best alpha is 0.9.

## Ridge Model with Best Alpha From Gridsearch

In [142]:
#Ridge model with alpha of 0.9
# fit a ridge regression model to the data
model_RG = Ridge(alpha = 0.9)
model_RG.fit(crime.features, crime.target)
print(model_RG)
# make predictions
expected_RG = crime.target
predicted_RG= model_RG.predict(crime.features)
# summarize the fit of the model
print("Coef", model_RG.intercept_, model_RG.coef_)
print("MSE", mean_squared_error(expected_RG, predicted_RG))

Ridge(alpha=0.9, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)
Coef 0.5173705618310256 [-0.03358301  0.02346494  0.18868294 -0.05095007 -0.0163454   0.04815463
  0.04442293 -0.18154648 -0.04562798  0.04300942 -0.06823997  0.03988346
  0.03216001 -0.12895388  0.03735831 -0.15014698  0.06112489  0.01433634
 -0.08141631  0.06826258 -0.01857973 -0.17129524 -0.02028281 -0.03072882
  0.02466351  0.04558191  0.03816789 -0.14098977 -0.0921011   0.05214113
  0.0489033  -0.01679304  0.15970447 -0.05631288 -0.01740876  0.0631866
  0.0529413   0.16295859  0.14185528 -0.08927749 -0.03236113  0.02386911
 -0.06242945 -0.1871576  -0.04153466 -0.00820907  0.03450358 -0.15055241
 -0.06280419  0.13778249 -0.09940188  0.02394293 -0.00927753 -0.01848045
  0.00711948 -0.02413413 -0.02040001  0.06906558  0.0115413  -0.01904156
 -0.11271422 -0.05497747 -0.05952522  0.21393035 -0.08109784 -0.06246989
 -0.12412988  0.17029867  0.07512338  0.0172

My final Ridge model I used the 0.9 alpha that was considered my best alpha based on the grid search and cross validation I did in the previous step.  This produced a MSE of 0.0168.  Based purely on the MSE, I would say that my best Ridge Regression model is my 3rd one with the alpha of 0.5.

# LASSO Regression

In [117]:
# Lasso Regression-Least Absolute Shrinkage and Selection Operator
from sklearn.linear_model import Lasso

## Basic LASSO Model

In [120]:
# fit a LASSO model to the data
model_LAS = Lasso(alpha=1)
model_LAS.fit(crime.features, crime.target)
print(model_LAS)
# make predictions
expected_LAS = crime.target
predicted_LAS = model_LAS.predict(crime.features)
# summarize the fit of the model
print("Coef", model_LAS.intercept_,model_LAS.coef_)
print("MSE", mean_squared_error(expected_LAS, predicted_LAS))

Lasso(alpha=1, copy_X=True, fit_intercept=True, max_iter=1000,
   normalize=False, positive=False, precompute=False, random_state=None,
   selection='cyclic', tol=0.0001, warm_start=False)
Coef 0.2379789368104313 [ 0. -0.  0. -0.  0.  0.  0.  0.  0.  0.  0.  0. -0. -0. -0. -0.  0.  0.
 -0. -0. -0. -0. -0. -0. -0. -0.  0.  0.  0.  0. -0.  0. -0. -0. -0.  0.
 -0.  0.  0.  0.  0.  0. -0. -0. -0. -0. -0. -0.  0.  0.  0.  0.  0.  0.
  0.  0.  0.  0.  0. -0.  0.  0.  0. -0. -0.  0. -0.  0.  0. -0.  0. -0.
 -0.  0.  0. -0.  0.  0. -0. -0. -0. -0. -0. -0. -0.  0.  0.  0.  0.  0.
  0. -0. -0.  0. -0.  0.  0.  0.  0.]
MSE 0.05425474178302209


The basic Lasso Regression model produced a MSE of 0.0542, with the alpha set to 1.

In [128]:
# Adjust the alpha to 0.01 from 1.0
model_LAS = Lasso(alpha=0.01)
model_LAS.fit(crime.features, crime.target)
print(model_LAS)
# make predictions
expected_LAS = crime.target
predicted_LAS = model_LAS.predict(crime.features)
# summarize the fit of the model
print("Coef", model_LAS.intercept_,model_LAS.coef_)
print("MSE", mean_squared_error(expected_LAS, predicted_LAS))

Lasso(alpha=0.01, copy_X=True, fit_intercept=True, max_iter=1000,
   normalize=False, positive=False, precompute=False, random_state=None,
   selection='cyclic', tol=0.0001, warm_start=False)
Coef 0.4588145567123556 [ 0.         -0.          0.         -0.17599284  0.          0.
 -0.         -0.         -0.          0.          0.          0.
 -0.         -0.         -0.         -0.          0.          0.
 -0.         -0.         -0.         -0.         -0.         -0.
 -0.         -0.          0.          0.          0.          0.
 -0.          0.         -0.         -0.         -0.          0.
 -0.          0.          0.          0.          0.          0.
 -0.         -0.23323255 -0.         -0.         -0.         -0.
  0.          0.22628628  0.          0.          0.          0.
  0.          0.          0.          0.          0.         -0.
  0.          0.          0.         -0.         -0.          0.
 -0.          0.          0.         -0.          0.         -0.
 -0.

For my second Lasso model, I adjusted my alpha down to 0.01 which caused my MSE to drop all the way down to 0.0238.

## Full Grid Search With 3 Different Grids

### Grid 1

In [130]:
param_grid = {"alpha": [0.05, 0.01, 0.005]}  #different penalties I am playing with

# run grid search
grid_search = GridSearchCV(model_LAS, param_grid=param_grid,n_jobs=-1)
grid_search.fit(crime.features, crime.target)
print("Best", grid_search.best_params_) 

Best {'alpha': 0.005}


For my first grid, I am feeding my Lasso model 3 alphas, including the alpha I used in my previous model that saw my MSE drop from the original model.  I also gave it an alpha that is above and below that value to see for my second grid where I need to adjust to next.  For grid one, my best alpha was 0.005.

### Grid 2

In [131]:
param_grid = {"alpha": [0.005, 0.001, 0.0005]}  #different penalties I am playing with

# run grid search
grid_search = GridSearchCV(model_LAS, param_grid=param_grid,n_jobs=-1)
grid_search.fit(crime.features, crime.target)
print("Best", grid_search.best_params_) 

Best {'alpha': 0.0005}


For my second grid, I again fed it 3 values including 0.005 from my last model.  This time, the grid search produced a result saying my best alpha is 0.0005, continuing our trend downwards in the alpha that will produce our lowest MSE.

### Grid 3

In [132]:
param_grid = {"alpha": [0.0005, 0.0001, 0.00005]}  #different penalties I am playing with

# run grid search
grid_search = GridSearchCV(model_LAS, param_grid=param_grid,n_jobs=-1)
grid_search.fit(crime.features, crime.target)
print("Best", grid_search.best_params_) 

Best {'alpha': 0.0001}


After running the grid search 3 different times, adjusting the alpha each time, I believe that using an alpha of 0.0001 will be best for my model.

## Final Lasso Model With Best Alpha

In [133]:
# Adjust the alpha to 0.0001 as indicated from the grid search
model_LAS = Lasso(alpha=0.0001)
model_LAS.fit(crime.features, crime.target)
print(model_LAS)
# make predictions
expected_LAS = crime.target
predicted_LAS = model_LAS.predict(crime.features)
# summarize the fit of the model
print("Coef", model_LAS.intercept_,model_LAS.coef_)
print("MSE", mean_squared_error(expected_LAS, predicted_LAS))

Lasso(alpha=0.0001, copy_X=True, fit_intercept=True, max_iter=1000,
   normalize=False, positive=False, precompute=False, random_state=None,
   selection='cyclic', tol=0.0001, warm_start=False)
Coef 0.5238139625025185 [-0.          0.          0.19049555 -0.03015929 -0.          0.03410396
  0.00672702 -0.16613386  0.          0.00404662 -0.03846941  0.03568629
  0.         -0.12763614  0.02844845 -0.11030302  0.03352877  0.
 -0.08169029  0.         -0.         -0.07935263 -0.01204218 -0.026911
  0.02399961  0.03932173 -0.         -0.13076952 -0.04032571  0.
  0.01614339 -0.0145753   0.0983234  -0.03352553  0.          0.01870041
  0.          0.129231    0.11623283 -0.05595778 -0.          0.
 -0.         -0.24644451 -0.03705816 -0.          0.         -0.11080404
 -0.01857227  0.15856061 -0.08152465  0.00188986 -0.         -0.
 -0.          0.          0.          0.01383976  0.         -0.
 -0.05862602 -0.03504505 -0.02771786  0.10355384 -0.00300917 -0.0054601
 -0.02941436  0.148906

For my final Lasso model, I used my best alpha of 0.0001 that was produced from my 3 separate grid searches and narrowing down my results.  For this model, that alpha value gives an output of a MSE of 0.01711 which is our best result from the Lasso model, although not as good as our best results from the Ridge models.

# ElasticNet Regression

In [143]:
# ElasticNet Regression
from sklearn.linear_model import ElasticNet

In [144]:
# Basic ElasticNet Regression Model
model_EN = ElasticNet(alpha=1)
model_EN.fit(crime.features, crime.target)
print(model_EN)
# make predictions
expected_EN = crime.target
predicted_EN = model_EN.predict(crime.features)
# summarize the fit of the model
print("Coef", model_EN.intercept_, model_EN.coef_)
print("MSE", mean_squared_error(expected_EN, predicted_EN))

ElasticNet(alpha=1, copy_X=True, fit_intercept=True, l1_ratio=0.5,
      max_iter=1000, normalize=False, positive=False, precompute=False,
      random_state=None, selection='cyclic', tol=0.0001, warm_start=False)
Coef 0.2379789368104313 [ 0. -0.  0. -0.  0.  0.  0.  0.  0.  0.  0.  0. -0. -0. -0. -0.  0.  0.
 -0. -0. -0. -0. -0. -0. -0. -0.  0.  0.  0.  0. -0.  0. -0. -0. -0.  0.
 -0.  0.  0.  0.  0.  0. -0. -0. -0. -0. -0. -0.  0.  0.  0.  0.  0.  0.
  0.  0.  0.  0.  0. -0.  0.  0.  0. -0. -0.  0. -0.  0.  0. -0.  0. -0.
 -0.  0.  0. -0.  0.  0. -0. -0. -0. -0. -0. -0. -0.  0.  0.  0.  0.  0.
  0. -0. -0.  0. -0.  0.  0.  0.  0.]
MSE 0.05425474178302209


My first ElasticNet model used the default values, which means an alpha of 1.  This produced an MSE of 0.0542.

In [146]:
# ElasticNet Regression Model adjusting Alpha from 1 to 0.5
model_EN = ElasticNet(alpha=0.5)
model_EN.fit(crime.features, crime.target)
print(model_EN)
# make predictions
expected_EN = crime.target
predicted_EN = model_EN.predict(crime.features)
# summarize the fit of the model
print("Coef", model_EN.intercept_, model_EN.coef_)
print("MSE", mean_squared_error(expected_EN, predicted_EN))

ElasticNet(alpha=0.5, copy_X=True, fit_intercept=True, l1_ratio=0.5,
      max_iter=1000, normalize=False, positive=False, precompute=False,
      random_state=None, selection='cyclic', tol=0.0001, warm_start=False)
Coef 0.2379789368104313 [ 0. -0.  0. -0.  0.  0.  0.  0.  0.  0.  0.  0. -0. -0. -0. -0.  0.  0.
 -0. -0. -0. -0. -0. -0. -0. -0.  0.  0.  0.  0. -0.  0. -0. -0. -0.  0.
 -0.  0.  0.  0.  0.  0. -0. -0. -0. -0. -0. -0.  0.  0.  0.  0.  0.  0.
  0.  0.  0.  0.  0. -0.  0.  0.  0. -0. -0.  0. -0.  0.  0. -0.  0. -0.
 -0.  0.  0. -0.  0.  0. -0. -0. -0. -0. -0. -0. -0.  0.  0.  0.  0.  0.
  0. -0. -0.  0. -0.  0.  0.  0.  0.]
MSE 0.05425474178302209


My second ElasticNet model adjusted the alpha down to 0.5 which actually gave the exact same MSE result as the model using the default values.  I will need to make greater adjustments to my alpha either up or down to see my MSE adjust as well.

In [147]:
# ElasticNet Regression Model adjusting Alpha from 0.5 to 0.05
model_EN = ElasticNet(alpha=0.05)
model_EN.fit(crime.features, crime.target)
print(model_EN)
# make predictions
expected_EN = crime.target
predicted_EN = model_EN.predict(crime.features)
# summarize the fit of the model
print("Coef", model_EN.intercept_, model_EN.coef_)
print("MSE", mean_squared_error(expected_EN, predicted_EN))

ElasticNet(alpha=0.05, copy_X=True, fit_intercept=True, l1_ratio=0.5,
      max_iter=1000, normalize=False, positive=False, precompute=False,
      random_state=None, selection='cyclic', tol=0.0001, warm_start=False)
Coef 0.2888739692310527 [ 0.         -0.          0.02051874 -0.0798286   0.          0.
  0.          0.          0.          0.          0.          0.
 -0.         -0.         -0.         -0.          0.          0.
 -0.         -0.         -0.         -0.         -0.         -0.
 -0.         -0.          0.          0.          0.          0.
 -0.          0.         -0.         -0.         -0.          0.
 -0.          0.          0.          0.          0.          0.
 -0.         -0.03500081 -0.         -0.         -0.         -0.
  0.          0.10924545  0.          0.          0.          0.
  0.          0.          0.          0.          0.         -0.
  0.          0.          0.         -0.         -0.          0.
 -0.          0.          0.         -0.    

My 3rd model used an alpha of 0.05, which finally saw my MSE go down to 0.0382.  This did lower the MSE, but I will next use a grid search combined with cross validation to see if continued dropping of the alpha can bring my MSE even lower.

## Grid Search to narrow down alpha with Cross Validation 5 Times

In [149]:
# use a full grid over several parameters
param_grid = {"alpha": [0.00005, 0.0001, 0.0005, 0.001, 0.005, 0.01, 0.05, 0.1, 0.25]}  #different penalties I am playing with

# run grid search
grid_search = GridSearchCV(model_EN, param_grid=param_grid,n_jobs=-1, cv = 5)
grid_search.fit(crime.features, crime.target)
print("Best", grid_search.best_params_) 

Best {'alpha': 0.0001}


Grid search between 9 alpha values and a cross validation being performed 5 times yields a best alpha result of 0.0001.  My grid search included values below and above the previously used alpha of 0.05.  The cross validation allowed each of those alphas to run through the model 5 times using separate samples of data.  The combination of the alphas in the grid search and the cross validation produced a best alpha of 0.0001.

In [150]:
# ElasticNet Regression Model adjusting Alpha to 0.0001 as indicated as the best alpha for the model
model_EN = ElasticNet(alpha=0.0001)
model_EN.fit(crime.features, crime.target)
print(model_EN)
# make predictions
expected_EN = crime.target
predicted_EN = model_EN.predict(crime.features)
# summarize the fit of the model
print("Coef", model_EN.intercept_, model_EN.coef_)
print("MSE", mean_squared_error(expected_EN, predicted_EN))

ElasticNet(alpha=0.0001, copy_X=True, fit_intercept=True, l1_ratio=0.5,
      max_iter=1000, normalize=False, positive=False, precompute=False,
      random_state=None, selection='cyclic', tol=0.0001, warm_start=False)
Coef 0.5502590618780592 [-0.          0.          0.19507642 -0.03512122 -0.00218166  0.05111446
  0.02543829 -0.23239891 -0.          0.0164314  -0.07084361  0.03857591
  0.         -0.16064697  0.03645328 -0.1457448   0.05110971  0.
 -0.08678994  0.07205893  0.         -0.16377285 -0.01772601 -0.0295398
  0.02360377  0.04424481  0.         -0.14353103 -0.06335331  0.01600431
  0.04089252 -0.01043003  0.15378626 -0.04713185 -0.          0.04821186
  0.02795142  0.15592205  0.1537795  -0.1011234  -0.         -0.
 -0.         -0.2651728  -0.03577887 -0.          0.02704295 -0.14579233
 -0.03897397  0.14191253 -0.09909447  0.00877475 -0.00314502 -0.00255198
 -0.         -0.         -0.          0.01699657  0.         -0.00441094
 -0.1007513  -0.04476726 -0.06181035  0.3007

My final version of the ElasticNet model used my best alpha from above, 0.0001.  This produced my lowest MSE for my ElasticNet Regressions of 0.0168 which is very close to the MSE we got in our best Ridge Regression model.

# SGD Regressor

In [151]:
# SGD Regressor
from sklearn.linear_model import SGDRegressor

In [159]:
# Basic SGD Regressor Model, l2 penalty 
model_SGD = SGDRegressor()
model_SGD.fit(crime.features, crime.target)
print(model_SGD)
# make predictions
expected_SGD = crime.target
predicted_SGD = model_SGD.predict(crime.features)
# summarize the fit of the model
print("Coef", model_SGD.intercept_, model_SGD.coef_)
print("MSE", mean_squared_error(expected_SGD, predicted_SGD))

SGDRegressor(alpha=0.0001, average=False, epsilon=0.1, eta0=0.01,
       fit_intercept=True, l1_ratio=0.15, learning_rate='invscaling',
       loss='squared_loss', max_iter=None, n_iter=None, penalty='l2',
       power_t=0.25, random_state=None, shuffle=True, tol=None, verbose=0,
       warm_start=False)
Coef [0.02572085] [ 2.23021108e-02  1.62687962e-03  1.11497164e-01 -6.71746869e-02
 -2.89160519e-04  1.01098986e-02 -4.13799898e-03 -2.36503460e-03
 -8.84745011e-03  1.95581189e-02  2.25192280e-02  4.85209155e-02
 -1.05863535e-03 -5.19274640e-03 -7.79490051e-03 -2.10642421e-02
  1.80020351e-02  4.03548469e-02  6.06828664e-03 -3.52423516e-03
  6.88668834e-03  2.18661364e-02 -3.91562959e-03  1.00807534e-03
  1.77244073e-02  1.56768514e-02  2.56549478e-02  2.73409891e-02
  9.12273702e-03  2.60686998e-02 -1.98586122e-03  2.53747178e-02
  2.58140619e-03 -9.33856541e-03  1.44714709e-03  1.10690433e-02
  4.35567584e-03  6.10836145e-02  1.67243916e-02  5.42480889e-02
  5.94458399e-02  1.294045

C:\Users\JM025575\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDRegressor'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


My first SGD Regressor model using all of the default parameter values, produced an MSE of 0.01978.

In [155]:
# SGD Regressor Model, adjusting penalty to l1 and alpha to 0.01
model_SGD = SGDRegressor(alpha = 0.01, penalty = 'l1')
model_SGD.fit(crime.features, crime.target)
print(model_SGD)
# make predictions
expected_SGD = crime.target
predicted_SGD = model_SGD.predict(crime.features)
# summarize the fit of the model
print("Coef", model_SGD.intercept_, model_SGD.coef_)
print("MSE", mean_squared_error(expected_SGD, predicted_SGD))

SGDRegressor(alpha=0.01, average=False, epsilon=0.1, eta0=0.01,
       fit_intercept=True, l1_ratio=0.15, learning_rate='invscaling',
       loss='squared_loss', max_iter=None, n_iter=None, penalty='l1',
       power_t=0.25, random_state=None, shuffle=True, tol=None, verbose=0,
       warm_start=False)
Coef [0.0876114] [ 0.          0.          0.11236818 -0.04802726  0.          0.
  0.          0.          0.          0.          0.          0.02716932
  0.          0.          0.          0.          0.          0.05405186
  0.          0.          0.          0.          0.          0.
  0.          0.          0.          0.03204841  0.          0.01641043
  0.          0.01931893  0.          0.          0.          0.
  0.          0.04121438  0.          0.04328078  0.04988087  0.
 -0.0081412  -0.02309844 -0.00303249  0.          0.          0.
  0.          0.13249219  0.          0.          0.          0.
  0.          0.          0.          0.          0.          0.
  0. 

C:\Users\JM025575\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDRegressor'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


For my second model, I adjusted my alpha up to 0.01 from the default of 0.0001 and changed my penalty from the default L2 to L1. These changes caused my MSE to go up from the basic model to 0.0247.

In [164]:
# SGD Regressor Model, adjusting penalty to l1 and alpha to 0.00001
model_SGD = SGDRegressor(alpha = 0.00001, penalty = 'l1')
model_SGD.fit(crime.features, crime.target)
print(model_SGD)
# make predictions
expected_SGD = crime.target
predicted_SGD = model_SGD.predict(crime.features)
# summarize the fit of the model
print("Coef", model_SGD.intercept_, model_SGD.coef_)
print("MSE", mean_squared_error(expected_SGD, predicted_SGD))

SGDRegressor(alpha=1e-05, average=False, epsilon=0.1, eta0=0.01,
       fit_intercept=True, l1_ratio=0.15, learning_rate='invscaling',
       loss='squared_loss', max_iter=None, n_iter=None, penalty='l1',
       power_t=0.25, random_state=None, shuffle=True, tol=None, verbose=0,
       warm_start=False)
Coef [0.02440499] [ 2.36662663e-02  1.08056787e-03  1.11079553e-01 -6.87879884e-02
  1.21457503e-03  7.67220874e-03 -3.86285276e-03 -2.32593748e-03
 -8.26058885e-03  1.92331425e-02  2.40447176e-02  4.70683271e-02
 -2.16559131e-03 -6.31983309e-03 -6.94729847e-03 -2.24211884e-02
  1.77775977e-02  4.12271686e-02  4.93417367e-03 -4.77494412e-03
  6.34105458e-03  2.14765393e-02 -4.51148688e-03 -3.67035540e-04
  1.78466866e-02  1.39448773e-02  2.62127535e-02  2.72325363e-02
  8.65799123e-03  2.58277373e-02 -2.30642567e-03  2.34518541e-02
  1.28593091e-03 -9.83142883e-03  1.58301499e-03  9.65087739e-03
  4.14385044e-03  5.96674274e-02  1.71410698e-02  5.40028174e-02
  5.87026553e-02  1.2405936

C:\Users\JM025575\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDRegressor'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


My third model kept my penalty at L1, but adjusted the alpha below the default value to 0.00001.  This lowered my MSE to 0.01909, which was my best SGD model yet.

## Grid Search to narrow down alpha with Cross Validation 10 Times

In [165]:
# use a full grid over several parameters
param_grid = {"penalty": ['none', 'l2', 'l1', 'elasticnet']}  #different penalties I am playing with

# run grid search
grid_search = GridSearchCV(model_SGD, param_grid=param_grid,n_jobs=-1, cv = 10)
grid_search.fit(crime.features, crime.target)
print("Best", grid_search.best_params_) 

Best {'penalty': 'elasticnet'}


C:\Users\JM025575\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDRegressor'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


Next, I used a grid search using the 4 possible penalty values and a cross validation ten times to see which of the penalty values is best to use.  My result was that the best penalty to use is 'elasticnet'.

In [168]:
# SGD Regressor Model, adjusting penalty 'none' based on the grid search results and cross validation
model_SGD = SGDRegressor(alpha = 0.00001, penalty = 'elasticnet')
model_SGD.fit(crime.features, crime.target)
print(model_SGD)
# make predictions
expected_SGD = crime.target
predicted_SGD = model_SGD.predict(crime.features)
# summarize the fit of the model
print("Coef", model_SGD.intercept_, model_SGD.coef_)
print("MSE", mean_squared_error(expected_SGD, predicted_SGD))

SGDRegressor(alpha=1e-05, average=False, epsilon=0.1, eta0=0.01,
       fit_intercept=True, l1_ratio=0.15, learning_rate='invscaling',
       loss='squared_loss', max_iter=None, n_iter=None,
       penalty='elasticnet', power_t=0.25, random_state=None, shuffle=True,
       tol=None, verbose=0, warm_start=False)
Coef [0.02359935] [ 0.02253122  0.0006861   0.11079761 -0.0680806   0.00065694  0.00768694
 -0.00568697 -0.00341933 -0.00947525  0.01953771  0.02283487  0.04834698
 -0.00217972 -0.00653952 -0.00813939 -0.02157164  0.01813252  0.04064194
  0.00511009 -0.004651    0.00643762  0.02174853 -0.00408055  0.00046418
  0.01807431  0.01498825  0.0256437   0.02617906  0.00867522  0.02562878
 -0.00288228  0.02205893  0.00180403 -0.01245598  0.00215429  0.00899576
  0.00342779  0.06081988  0.0159269   0.05354806  0.05891996  0.01214231
 -0.0465784  -0.05368277 -0.04760003 -0.04109479  0.00975825 -0.00073252
  0.02766847  0.09522178  0.00870417  0.00136837  0.00146916  0.00650692
  0.01348518

C:\Users\JM025575\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDRegressor'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


Finally, I used the result from my grid search and cross validation and paired that with the lowest alpha I had used and ended up getting my lowest MSE yet with this model, 0.01907.

# Hold for explanation of what best model is and why

After reviewing my 4 different Ridge, Lasso, ElasticNet, and SGD Regressor, the best model I would say that my best model was my third Ridge Regression model, where I used an alpha of 0.5 and got a MSE of 0.0167.  I got various other MSE values that came close to that after tuning those models in various ways and using grid search and cross validation to more efficiently search for my best alpha and penalty values.  However, based solely on the lowest MSE, my Ridge Regression Model with the alpha of 0.5 is my best performing model.